stake-pool/program/src/processor.rs

```diff
@@ -430,79 +691,384 @@ impl Processor {
+        // Stake account to split
+        let stake_split_from = next_account_info(account_info_iter)?;
+        // Unitialized stake account to receive withdrawal
+        let stake_split_to = next_account_info(account_info_iter)?;
+        // User account to set as a new withdraw authority
+        let user_stake_authority = next_account_info(account_info_iter)?;
+        // User account with pool tokens to burn from
+        let burn_from_info = next_account_info(account_info_iter)?;
+        // Pool token mint account
+        let pool_mint_info = next_account_info(account_info_iter)?;
+        // Clock sysvar account
+        let clock_info = next_account_info(account_info_iter)?;
+        let clock = &Clock::from_account_info(clock_info)?;
+        // Pool token program id
+        let token_program_info = next_account_info(account_info_iter)?;
+        // Stake program id
+        let stake_program_info = next_account_info(account_info_iter)?;
+
+        // Check program ids
+        if *stake_program_info.key != stake::id() {
+            return Err(ProgramError::IncorrectProgramId);
         }
+
+        let mut stake_pool = State::deserialize(&stake_pool_info.data.borrow())?.stake_pool()?;
+
+        // Check authority account
+        stake_pool.check_authority_withdraw(withdraw_info.key, program_id, stake_pool_info.key)?;
+
         if stake_pool.token_program_id != *token_program_info.key {
-            return Err(Error::InvalidInput.into());
+            return Err(StakePoolError::InvalidProgramAddress.into());
+        }
+
+        // Check validator stake account list storage
+        if *validator_stake_list_info.key != stake_pool.validator_stake_list {
+            return Err(StakePoolError::InvalidValidatorStakeList.into());
+        }
+
+        // Check stake pool last update epoch
+        if stake_pool.last_update_epoch < clock.epoch {
+            return Err(StakePoolError::StakeListAndPoolOutOfDate.into());
         }
 
-        let stake_amount = **stake_to_claim.lamports.borrow();
+        // Read validator stake list account and check if it is valid
+        let mut validator_stake_list =
+            ValidatorStakeList::deserialize(&validator_stake_list_info.data.borrow())?;
+        if !validator_stake_list.is_initialized {
+            return Err(StakePoolError::InvalidState.into());
+        }
+
+        let validator_account =
+            Self::get_validator_checked(program_id, stake_pool_info, stake_split_from)?;
+
+        let validator_list_item = validator_stake_list
+            .find_mut(&validator_account)
+            .ok_or(StakePoolError::ValidatorNotFound)?;
+
         let pool_amount = stake_pool
             .calc_pool_withdraw_amount(stake_amount)
-            .ok_or(Error::CalculationFailure)?;
-        let pool_amount = <u64>::try_from(pool_amount).or(Err(Error::CalculationFailure))?;
+            .ok_or(StakePoolError::CalculationFailure)?;
+
+        Self::stake_split(
```

Comment:
I looked at the stake program, and I think we should be able to trust the program when doing the split.  There might be some errors if people don't withdraw enough, in which case they need to prefund the destination stake account.  All good!